scheds/rust/scx_flash/src/bpf/main.bpf.c

```diff
@@ -818,29 +818,21 @@ void BPF_STRUCT_OPS(flash_enqueue, struct task_struct *p, u64 enq_flags)
 void BPF_STRUCT_OPS(flash_dispatch, s32 cpu, struct task_struct *prev)
 {
 	/*
-	 * If the task can still run and it's holding a user-space lock, let it
-	 * run for another round.
+	 * If the task can still run, let it run for another round on the
+	 * same CPU. Usually these are the tasks which we want to prioritize
+	 * (i.e. shorter runtime), so let it finish earlier leads to better
+	 * latency outcome.
 	 */
-	if (prev && (prev->scx.flags & SCX_TASK_QUEUED) &&
-	    is_task_locked(prev)) {
-		task_unlock(prev);
+	if (prev && (prev->scx.flags & SCX_TASK_QUEUED)) {
+		if (is_task_locked(prev))
+			task_unlock(prev);
 		task_refill_slice(prev);
```

Comment:
Hm.. but in this way we are refilling the time slice for all the tasks that still want to run. Shouldn't we do this only if the task has a lock?  Basically moving the refill inside is_task_locked()

Reply:
Yeah I agree, I should change this part and test it again.
My original intent was more aggressive, as long as the task can still run we let it run one more time (on an assumption that most task still has `slice > 0` wouldn't run for too long.) But the assumption indeed is too optimistic and too aggressive.

Though I have a question, what kind of task would be holding a user-space lock ? does it indicate some process calling `pthread_mutex_lock()` or something like that ?

Reply:
The idea here is to trace when user-space tasks are holding a futex. If a task holding a futex uses all its assigned time slice it'd be de-scheduled in favor of another task, but the other task may still need to acquire the same futex, introducing a priority-inversion issue.

So, the idea is to trace when a task is holding a futex and if its time slice expires give it another time slice "for free", so that it can (hopefully) complete its critical section and release the lock. Of course we also want to prevent tasks from abusing this "bonus", so when we give another "free" timeslice we also mark the task a "unlocked".